crates/evm/fuzz/src/strategies/state.rs

```diff
@@ -36,12 +37,31 @@ impl EvmFuzzState {
         }
     }
 
-    /// Collects state changes from a [StateChangeset] and logs into an [EvmFuzzState] according to
-    /// the given [FuzzDictionaryConfig].
-    pub fn collect_state_from_call(&self, logs: &[Log], state_changeset: &StateChangeset) {
+    /// Collects call result (if any), state changes from a [StateChangeset] and logs into an
+    /// [EvmFuzzState] according to the given [FuzzDictionaryConfig].
+    pub fn collect_state_from_call(
+        &self,
+        function: &Option<Function>,
+        result: &Bytes,
+        logs: &[Log],
+        state_changeset: &StateChangeset,
+        run_depth: u32,
+    ) {
         let mut dict = self.inner.write();
 
-        // Insert log topics and data.
+        match function {
+            Some(func) => {
+                // Decode result and collect samples to be used in subsequent fuzz runs.
+                if !result.is_empty() {
+                    if let Ok(decoded_result) = func.abi_decode_output(result, true) {
+                        dict.insert_sample_value(decoded_result, run_depth);
```

Comment:
what's the motivation behind using `run_depth` as a limit for values of a certain type?

Reply:
We need a limit for samples to remain relevant, if it is too low then we're going to use same values / repeat same test too many times, at the same time if it grows too big then samples won't be exercised enough to reveal failures due to potential dependencies.
The test depth is a limit I come up to make sure that, on the extreme case where we run with test depth of `n` for a fuzz target containing same `n` functions with one return value, we collect one sample from each function.
We could also introduce a new config for such to replace default behavior.

Reply:
> The test depth is a limit I come up to make sure that, on the extreme case where we run with test depth of `n` for a fuzz target containing same `n` functions with one return value, we collect one sample from each function.

Is this assumption true given how fuzz targets and selectors are chosen? i.e. I don't think there's a guarantee you cover all `n` functions with a depth of `n`, especially if there's >1 contract. Also related to https://github.com/foundry-rs/foundry/issues/2986

But regardless, I'll echo my thoughts in https://github.com/foundry-rs/foundry/pull/7666#issuecomment-2078416499 — performance-wise (catching DSChief) this PR looks really good, and I think it's better to ship as-is and revisit this limit once we have benchmarks, than to block this PR on deciding a good value. Since before benchmarks we're just taking our best guess at what sensible values are anyway :)

Reply:
> Is this assumption true given how fuzz targets and selectors are chosen? i.e. I don't think there's a guarantee you cover all `n` functions with a depth of `n`, especially if there's >1 contract. Also related to #2986
> 

Indeed, that is not a guarantee just the ideal case, as you say it can be improved as we go. Only concern could be that if this limit is too low then other relevant data collected won't be exercised enough.